test-framework/junit5/src/main/java/io/quarkus/test/junit/AbstractJvmQuarkusTestExtension.java

```diff
@@ -57,106 +46,70 @@ public class AbstractJvmQuarkusTestExtension extends AbstractQuarkusTestWithCont
-                    }
-                    for (SourceDir src : module.getMainSources().getResourceDirs()) {
-                        addToBuilderIfConditionMet.accept(src.getOutputDir());
-                    }
-                    break;
-                }
-            }
-            if (testClassesDir == null) {
-                final StringBuilder sb = new StringBuilder();
-                sb.append("Failed to locate ").append(requiredTestClass.getName()).append(" in ");
-                for (String classifier : module.getSourceClassifiers()) {
-                    final ArtifactSources sources = module.getSources(classifier);
-                    if (sources.isOutputAvailable()) {
-                        for (SourceDir d : sources.getSourceDirs()) {
-                            if (Files.exists(d.getOutputDir())) {
-                                sb.append(System.lineSeparator()).append(d.getOutputDir());
-                            }
-                        }
-                    }
-                }
-                throw new RuntimeException(sb.toString());
-            }
-            testClassLocation = testClassesDir;
-
-        } else {
-            if (System.getProperty(BootstrapConstants.OUTPUT_SOURCES_DIR) != null) {
-                final String[] sourceDirectories = System.getProperty(BootstrapConstants.OUTPUT_SOURCES_DIR).split(",");
-                for (String sourceDirectory : sourceDirectories) {
-                    final Path directory = Paths.get(sourceDirectory);
-                    addToBuilderIfConditionMet.accept(directory);
-                }
-            }
 
-            testClassLocation = getTestClassesLocation(requiredTestClass);
-            appClassLocation = getAppClassLocationForTestLocation(testClassLocation);
-            if (!appClassLocation.equals(testClassLocation)) {
-                addToBuilderIfConditionMet.accept(testClassLocation);
-                // if test classes is a dir, we should also check whether test resources dir exists as a separate dir (gradle)
-                // TODO: this whole app/test path resolution logic is pretty dumb, it needs be re-worked using proper workspace discovery
-                final Path testResourcesLocation = PathTestHelper.getResourcesForClassesDirOrNull(testClassLocation, "test");
-                addToBuilderIfConditionMet.accept(testResourcesLocation);
-            }
+        CuratedApplication curatedApplication = getCuratedApplication(requiredTestClass, context, shutdownTasks);
 
-            addToBuilderIfConditionMet.accept(appClassLocation);
-            final Path appResourcesLocation = PathTestHelper.getResourcesForClassesDirOrNull(appClassLocation, "main");
-            addToBuilderIfConditionMet.accept(appResourcesLocation);
-        }
+        // TODO need to handle the gradle case - can we put it in that method?
+        Path testClassLocation = getTestClassesLocation(requiredTestClass);
```

Comment:
```suggestion
        Path testClassLocation = getTestClassesLocation(requiredTestClass, curatedApplication);
```
Should it be that now?

Reply:
Yes, it probably could be. Sometimes the version that accepts the curated application has to fall back to the 'low tech' version, but it should be safe to call the smarter version first. Will give it a try.